[Fuchsia] Move physical shape layer compositing to Flutter

mikejurka opened the pull request:
I split the change up into smaller CLs. I haven't independently tested the sub-CLs, but just the full stack. No known issues at this time (the one issue I ran into today turned out to be a Scenic/Escher issue).

mikejurka commented on flow/scene_update_context.cc:
```diff
@@ -72,14 +73,13 @@ void SceneUpdateContext::CreateFrame(scenic::EntityNode entity_node,
-      rrect.radii(SkRRect::kUpperRight_Corner).x(),  // top_right_radius
-      rrect.radii(SkRRect::kLowerRight_Corner).x(),  // bottom_right_radius
-      rrect.radii(SkRRect::kLowerLeft_Corner).x()    // bottom_left_radius
+  scenic::RoundedRectangle shape(session_,        // session
```
This should just be a scenic::Rectangle

liyuqian replied:
If it's just a rectangle, why are we not using `scenic::Rectangle` here? Does `scenic::Rectangle` exist? It might be nice to write the answers as code comment just in case that someone will have similar questions by reading the code.

mikejurka replied:
yeah, my comment was meant to say I should just use scenic::Rectangle :) will include that in fixes

mikejurka replied:
Done

mikejurka commented on flow/scene_update_context.cc:
```diff
@@ -313,7 +313,11 @@ SceneUpdateContext::Frame::Frame(SceneUpdateContext& context,
+  // if there is a OpacityNode higher in the tree with opacity != 1. For now,
+  // clamp to a infinitesimally smaller value than 1, which does not cause visual
+  // problems in practice.
+  opacity_node_.SetOpacity(std::min(1 - FLT_EPSILON, opacity_ / 255.0f));
```
add Scenic bug for API for marking surfaces as transparent

liyuqian commented on flow/layers/physical_shape_layer.h:
```diff
@@ -37,11 +37,18 @@ class PhysicalShapeLayer : public ContainerLayer {
 #if defined(OS_FUCHSIA)
   void UpdateScene(SceneUpdateContext& context) override;
+
+  void UpdateSceneChildren(SceneUpdateContext& context);
```
I just realized that `UpdateSceneChildren` is `protected` in other layers but `public` in here. Shall we make them consistent as I don't see why this needs to be public. Also, it might be nice to add some documentation to these function declarations. Although we lack the documentation traditionally (just like tests), we're pushing for better documentations and tests for all new code.

arbreng replied:
You can avoid overriding it entirely if you follow one of my suggestions above

mikejurka replied:
Removed the separate method and moved inline, which removed a member variable at the cost of a little extra duplication.

liyuqian commented on flow/scene_update_context.cc:
```diff
@@ -222,6 +222,9 @@ SceneUpdateContext::ExecutePaintTasks(CompositorContext::ScopedFrame& frame) {
   }
   paint_tasks_.clear();
+  alpha_ = 1.f;
+  topmost_global_scenic_elevation_ = 10.f;
```
I curious where this `10.f ` comes from, and whether it's configurable.

mikejurka replied:
will pull out to constant

mikejurka replied:
Done

liyuqian commented on flow/scene_update_context.cc:
```diff
@@ -303,23 +310,14 @@ SceneUpdateContext::Frame::Frame(SceneUpdateContext& context,
   entity_node().AddChild(opacity_node_);
-  opacity_node_.SetOpacity(opacity_ / 255.0f);
+  // Scenic currently lacks an API to enable rendering of alpha channel; this only happens
+  // if there is a OpacityNode higher in the tree with opacity != 1. For now,
```
Do you mean "this only happens ... with opacity == 1"? Clamping with 1 - epsilon seems to avoid `opacity == 1` case instead of `opacity != 1` case.

mikejurka replied:
Re-read it, the original language should be accurate. Tried to re-word it to make it clearer.

liyuqian commented on flow/scene_update_context.h:
```diff
@@ -176,10 +175,25 @@ class SceneUpdateContext {
+
+  float GetGlobalElevationForNextScenicLayer() {
+    float elevation = topmost_global_scenic_elevation_;
+    topmost_global_scenic_elevation_ += 10.f;
```
same for the `10.f` here.

mikejurka replied:
Done.

liyuqian commented on shell/platform/fuchsia/flutter/vulkan_surface.h:
```diff
@@ -134,9 +134,9 @@ class VulkanSurface final
   // retained surface directly. Hence Flutter can't modify the surface during
   // retained rendering.
-  const scenic::EntityNode& GetRetainedNode() {
+  scenic::EntityNode* GetRetainedNode() {
```
The comment of this method says that we want to prevent Flutter from modifying the surface. Is changing `const scenic::EntityNode&` to `scenic::EntityNode*` breaking that promise?

mikejurka replied:
Yes. Updating comment.

arbreng commented on flow/layers/layer.h:
```diff
@@ -66,6 +66,11 @@ struct PrerollContext {
+#if defined(OS_FUCHSIA)
+  // True if, during the traversal so far, we have seen a child_scene_layer.
+  // Informs whether a layer needs to be system composited.
+  bool child_scene_layer_exists_below = false;
```
Why not use needs_system_composite() for this?  That is the intent of the flag, and with your other changes the *only* place it is set is inside of child_scene_layer.  It would be perfect for the job and you could avoid adding all of the logic related to this flag

mikejurka replied:
needs_system_composite() is for a particular layer, while this is defined for the *PrerollContext*. we need to track this value during a traversal, so physicalshapelayers can decide whether or not they should be system composited-- a big part of this change is that they usually are *not* system composited.

arbreng replied:
Yes, while needs_system_composite() is for a particular layer, ContainerLayer::PrerollChildren ensures that the needs_system_composite flags is propagated up the tree (or forward in the traversal order, if you like).

The means that every single PhysicalShapeLayer renderered before a ChildSceneLayer in the traversal order will already have the needs_system_composite() flag applied to it.

arbreng replied:
FWIW, I actually prefer your method of storing this in the context after thinking about this more....it prevents carrying a (mostly) useless bool inside of every single Layer.

I think you could also accomplish this by relying on your flag completely and deleting needs_system_composite().  The 2 flags are IMO redundant with one another, so one of them should go.

mikejurka replied:
I think you're right that it's propagated up the tree, but not that it's forward in the traversal order.


>That means that every single PhysicalShapeLayer renderered before a ChildSceneLayer in the traversal order will already have the needs_system_composite() flag applied to it, as long as you do the check after PrerollChildren() is called.

That's what we don't want actually. A PhysicalShapeLayer rendered before a ChildSceneLayer should not be system composited.

>I think you could also accomplish this by relying on your flag completely and deleting needs_system_composite(). The 2 flags are IMO redundant with one another, so one of them should go.

I think you're right, but there's a bit of risk there. At the very least it might lead to more calls to UpdateScene than are necessary. If it's ok, i'd prefer to skip entirely if this code goes away soon anyway, or at the very least defer

arbreng commented on flow/layers/child_scene_layer.cc:
```diff
@@ -20,6 +20,7 @@ ChildSceneLayer::ChildSceneLayer(zx_koid_t layer_id,
 void ChildSceneLayer::Preroll(PrerollContext* context, const SkMatrix& matrix) {
   TRACE_EVENT0("flutter", "ChildSceneLayer::Preroll");
   set_needs_system_composite(true);
+  context->child_scene_layer_exists_below = true;
```
Should be unneccesary; The line above handles this

mikejurka replied:
See above comment; we need both. child_scene_layer_exists_below is there so layers that are visited afterward will know that this was below them

mikejurka replied:
Removed children_need_system_compositing_.

arbreng commented on flow/layers/physical_shape_layer.cc:
```diff
@@ -78,30 +87,64 @@ void PhysicalShapeLayer::UpdateScene(SceneUpdateContext& context) {
+  // PhysicalShapeLayers that appear above the embedded content will be turned
+  // into their own Scenic layers.
+  if (child_layer_exists_below_) {
+    // Reset paint bounds, because shadows are not rendered in this case
```
Are you sure about this part?  There is no comment (or changes at all for that matter) in Paint()

arbreng replied:
Actually, after reading forward I'm even more sure.  If you follow my other comment below in this function, you can remove this

mikejurka replied:
Look for "// TODO: Re-enable shadow drawing here." in Pain().

I think our offline discussion clarified why we need to do a check for child_layer_exists_below_; keeping it. Let me know if that wasn't the case.

arbreng commented on flow/layers/physical_shape_layer.cc:
```diff
@@ -78,30 +87,64 @@ void PhysicalShapeLayer::UpdateScene(SceneUpdateContext& context) {
+                                    "flutter::PhysicalShapeLayer",
+                                    z_translation, this);
+
+    frame.AddPaintLayer(this);
```
If you loop through the children and add them, instead of adding this directly, you can avoid adding the #if that you did inside of Paint().

mikejurka replied:
Hmm I initially thought that this was necessary to draw the rounded rect border, but I might have been wrong. I can try again.

mikejurka replied:
We're going to want to re-add shadows soon; either we have an #if in paint, or just loop through the children. Is there a reason to strongly prefer one over the other? If not, I'll keep as-is.

arbreng replied:
Good point on shadows

arbreng commented on flow/layers/physical_shape_layer.h:
```diff
@@ -37,11 +37,18 @@ class PhysicalShapeLayer : public ContainerLayer {
 
  private:
+#if defined(OS_FUCHSIA)
+  float local_scenic_elevation_ = 0.0f;
```
#2 and #3 in this list should be unnecesary

mikejurka replied:
see comments above for child_layer_exists_below_

As far as children_need_system_compositing_ -- I think you are right, we can just get away with always calling UpdateSceneChildren in PhysicalShapeLayer::UpdateScene. I initially thought I needed it for correctness, but I think at this point it's just a tiny optimization. Getting rid of it will simplify things.

mikejurka replied:
got rid of #1 and #3 as per our discussion.

arbreng commented on flow/layers/physical_shape_layer.cc:
```diff
@@ -78,30 +87,64 @@ void PhysicalShapeLayer::UpdateScene(SceneUpdateContext& context) {
+  // If there is embedded Fuchsia content in the scene (a ChildSceneLayer),
+  // PhysicalShapeLayers that appear above the embedded content will be turned
+  // into their own Scenic layers.
+  if (child_layer_exists_below_) {
```
This will always be true, because if needs_system_composite() is true then child_layer_exists_below_ will be true by definition (now that you have deleted the other instances of it from OpacityLayer, etc).  This function is only ever called if needs_system_composite() is true so you can eliminate this if now.

arbreng commented on flow/layers/physical_shape_layer.cc:
```diff
@@ -78,30 +87,64 @@ void PhysicalShapeLayer::UpdateScene(SceneUpdateContext& context) {
+
+    // Node: UpdateSceneChildren needs to be called here so that |frame| is still
+    // in scope (and therefore alive) while UpdateSceneChildren is being called.
+    if (children_need_system_compositing_) {
```
You shouldn't need this if, UpdateSceneChildren will check internally whether each child needs system compositing or not and skip over it if it doesnt

mikejurka replied:
+1 I concur (see comment above)

mikejurka replied:
Removed.

arbreng commented on flow/layers/physical_shape_layer.cc:
```diff
@@ -78,30 +87,64 @@ void PhysicalShapeLayer::UpdateScene(SceneUpdateContext& context) {
+
+void PhysicalShapeLayer::UpdateSceneChildren(SceneUpdateContext& context) {
+  float scenic_elevation = context.scenic_elevation();
+  context.set_scenic_elevation(scenic_elevation + local_scenic_elevation_);
```
If you do this line in PhysicalShapeLayer::UpdateScene instead, you won't need to store `local_scenic_elevation_` as a member variable.

mikejurka replied:
i moved this because UpdateSceneChildren is called in two different places and this removed some duplicated code. But I think I can put the elevation logic at the very top and very bottom of UpdateScene which would eliminate the duplication and need for a variable, at the cost of a slight reduction in understandability.

mikejurka replied:
Removed local_scenic_elevation_.

arbreng commented on flow/layers/physical_shape_layer.h:
```diff
@@ -37,11 +37,18 @@ class PhysicalShapeLayer : public ContainerLayer {
 
  private:
+#if defined(OS_FUCHSIA)
+  float local_scenic_elevation_ = 0.0f;
```
Actually, #1 is also unnecesary if following one of my suggestions above

mikejurka replied:
+1

mikejurka replied:
Done.

arbreng commented on flow/layers/transform_layer.cc:
```diff
@@ -56,8 +56,12 @@ void TransformLayer::UpdateScene(SceneUpdateContext& context) {
 
-  SceneUpdateContext::Transform transform(context, transform_);
-  UpdateSceneChildren(context);
+  if (!transform_.isIdentity()) {
```
Nice!  this should cut down on the node count a good bit until we can roll out Embedder API

mikejurka replied:
Yep, it got rid of a few nodes. The changes to OpacityLayer removed even more.

arbreng commented on flow/layers/layer.h:
```diff
@@ -66,6 +66,11 @@ struct PrerollContext {
+#if defined(OS_FUCHSIA)
+  // True if, during the traversal so far, we have seen a child_scene_layer.
+  // Informs whether a layer needs to be system composited.
+  bool child_scene_layer_exists_below = false;
```
Mike: You can (and should) fix this bug as part of this CL by modifying the logic in ContainerLayer::Preroll to reset the needs_system_composite flag similar to the way it handles the has_platform_view flag

mikejurka replied:
I've been thinking about this more in terms of traversal order. Although all the layers are stored in a tree, there is (I assume) a deterministic traversal order, which defines the paint order.

Based on that, if child_scene_layer_exists_below is true, all physicalshapelayers painted afterward will be system composited.

that said, should performance overlay also be system composited if there is embedded content below?

arbreng replied:
Ah, I misread the bug a bit.  I agree, the proper fix is to system_composite the performance overlay if there is embedded content below.

mikejurka replied:
Discussed offline why we need to do things differently than has_platform_view.

Going to hold off any changes to performance overlay in this change since there's no regression (i.e. it's not popped up to a separate layer on ToT either).

mikejurka commented on flow/layers/child_scene_layer.cc:
```diff
@@ -49,7 +50,9 @@ void ChildSceneLayer::UpdateScene(SceneUpdateContext& context) {
 
-  view_holder->UpdateScene(context, offset_, size_, hit_testable_);
+  view_holder->UpdateScene(context, offset_, size_,
+                           sk_float_round2int(context.alphaf() * 255),
```
Done

mikejurka commented on flow/layers/physical_shape_layer.cc:
```diff
@@ -110,7 +153,19 @@ void PhysicalShapeLayer::Paint(PaintContext& context) const {
 
+
+#if defined(OS_FUCHSIA)
+  // TODO: Re-enable shadow drawing here.
```
Done

mikejurka commented:
Ok, added tests (see latest CL), and I also addressed comments in the last CL as well. Hope that's ok, I haven't found a productive workflow for going back and modifying the previous CLs.

liyuqian commented:
Thank you @mikejurka for adding the test! I wonder why all the presubmit tests are failing. Maybe you need to rebase against newest engine?

I also couldn't find my comment regarding https://github.com/flutter/flutter/issues/18320 although that issue showed that I referenced it from this PR...

Anyway, it mostly looks good to me, and the only thing obvious to me is that you forget one more `10.f` :)

mikejurka commented:
Replaced the stray  10.f with kScenicZElevationBetweenLayers
